bcc/symbol.go

```diff
@@ -0,0 +1,126 @@
+// Copyright 2016 PLUMgrid
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bcc
+
+import (
+	"fmt"
+	"regexp"
+	"sync"
+	"unsafe"
+)
+
+/*
+#cgo CFLAGS: -I/usr/include/bcc/compat
+#cgo LDFLAGS: -lbcc
+#include <bcc/bpf_common.h>
+#include <bcc/libbpf.h>
+#include <bcc/bcc_syms.h>
+extern void foreach_symbol_callback(char*, uint64_t);
+*/
+import "C"
+
+type symbolAddress struct {
+	name string
+	addr int64
+}
+
+//symbolCache will cache module lookups
+var symbolCache = struct {
+	cache         map[string][]*symbolAddress
+	currentModule string
+	lock          *sync.Mutex
+}{
+	cache:         map[string][]*symbolAddress{},
+	currentModule: "",
+	lock:          &sync.Mutex{},
+}
+
+type bccSymbol struct {
+	name         *C.char
+	demangleName *C.char
+	module       *C.char
+	offset       C.ulonglong
+}
+
+// resolveSymbolPath returns the file and offset to locate symname in module
+func resolveSymbolPath(module string, symname string, addr int64, pid int) (string, int64, error) {
+
+	symbol := &bccSymbol{}
+	symbolC := (*C.struct_bcc_symbol)(unsafe.Pointer(symbol))
+	moduleCS := C.CString(module)
+	defer C.free(unsafe.Pointer(moduleCS))
+	symnameCS := C.CString(symname)
+	defer C.free(unsafe.Pointer(symnameCS))
+
+	if pid == -1 {
+		pid = 0
```

Comment:
Yes, in module.go, in the various AttachUProbe functions, a pid of -1 signifies we wish to attach to all processes (the -1 eventually gets passed through to bpf_attach_uprobe). For e.g.:

```
// AttachUprobe attaches a uprobe fd to the symbol in the library or binary 'name'
// The 'name' argument can be given as either a full library path (/usr/lib/..),
// a library without the lib prefix, or as a binary with full path (/bin/bash)
// A pid can be given to attach to, or -1 to attach to all processes
func (bpf *Module) AttachUprobe(name, symbol string, fd, pid int) error {
	path, addr, err := resolveSymbolPath(name, symbol, 0x0, 0)
	if err != nil {
		return err
	}
	evName := fmt.Sprintf("p_%s_0x%x", uprobeRegexp.ReplaceAllString(path, "_"), addr)
	return bpf.attachUProbe(evName, BPF_PROBE_ENTRY, path, addr, fd, pid)
}
```

So we just switch it to be 0 here, meaning the symbol resolution will not be tied to any specific PID (I think).